src/bundle.rs

```diff
@@ -18,267 +18,505 @@ use crate::{
+
+cmake -S . -B build
+cmake --build build
+cp build/main .
+arbor-build-catalogue local cat
+./main $*
+".to_string())
+
+}
+
+fn mk_cmake() -> Result<String> {
+    Ok("
+cmake_minimum_required(VERSION 3.22)
+project(arbor-ring LANGUAGES CXX)
+set(CMAKE_CXX_STANDARD 17)
+
+find_package(arbor REQUIRED)
+find_package(nlohmann_json 3.11.0 REQUIRED)
+
+add_executable(main main.cxx)
+target_link_libraries(main PUBLIC arbor::arbor arbor::arborio arbor::arborenv nlohmann_json::nlohmann_json)
+".to_string())
+}
+
+fn mk_main_cxx() -> Result<String> {
+    Ok("
+#include <unordered_map>
+#include <vector>
+#include <string>
+#include <filesystem>
+#include <iostream>
+#include <fstream>
+#include <iterator>
+
+#include <arbor/context.hpp>
+#include <arbor/load_balance.hpp>
+#include <arbor/cable_cell.hpp>
+#include <arbor/simulation.hpp>
+#include <arbor/recipe.hpp>
+
+#include <arborio/neuroml.hpp>
+#include <arborio/cableio.hpp>
+
+#include <nlohmann/json.hpp>
+using json = nlohmann::json;
+
+arb::mechanism_catalogue compile(const std::filesystem::path& here) {
+    auto fn = here / \"local-catalogue.so\";
+    return arb::load_catalogue(fn);
+}
+
+std::string read_file(const std::filesystem::path& fn) {
+    auto fd = std::ifstream{fn};
+    return std::string(std::istreambuf_iterator<char>{fd}, {});
+}
 
-    let mut i_clamps = String::from("{");
-    let mut poisson = String::from("{");
-    let mut regular = String::from("{");
-    for (lbl, stimulus) in stimuli {
-        match stimulus {
-            Input::Pulse(delay, dt, stop) => {
-                i_clamps.push_str(&format!("'{lbl}': ({delay}, {dt}, {stop}), "))
+
+arb::locset on_segment(size_t seg, const std::string& frac) {
+    return arb::ls::on_components(std::stod(frac), arb::reg::named(std::to_string(seg)));
+};
+
+std::string mk_label(const std::string& pfx, size_t seg, const std::string& frac) {
+    return pfx + \"@seg_\" + std::to_string(seg) + \"_frac_\" + frac;
+}
+
+arb::cell_local_label_type mk_lid(const std::string& l) {
+     return {l, arb::lid_selection_policy::round_robin};
+}
+
+arb::cell_global_label_type mk_gid(arb::cell_gid_type c, const std::string& l) {
+     return {c, {l, arb::lid_selection_policy::round_robin}};
+}
+
+struct recipe: public arb::recipe {
+
+    recipe(const std::filesystem::path& cwd, const std::string& network): here{cwd} {
+        auto cat = compile(here);
+        gprop.default_parameters = arb::neuron_parameter_defaults;
+        gprop.catalogue.import(cat, prefix);
+
+        std::ifstream fd{(here / \"dat\" / network).replace_extension(\"json\")};
+        auto data = json::parse(fd);
+
+        gid_to_cell = data[\"gid_to_cell\"];
+        cell_to_morph = data[\"cell_to_morph\"];
+        i_clamps = data[\"i_clamps\"];
+        poisson_generators = data[\"poisson_generators\"];
+        regular_generators = data[\"regular_generators\"];
+        for (const auto& [k, v]: data[\"gid_to_inputs\"].items()) gid_to_inputs[std::stoi(k)] = v;
```

Comment:
Moving to external file will also help a lot with these escape characters and syntax highlighting. Could also become another PR of course.

Note, you could also do this:

```rust
let rawstr = r#"poisson_generators = data["poisson_generators"];"#;
```

from https://doc.rust-lang.org/rust-by-example/std/str.html

Reply:
Yeah, that raw string just occurred to me as well ;)